lib/cli/src/Cardano/CLI.hs

```diff
@@ -587,18 +611,58 @@ newCliKeyScheme = \case
     ledgerKeyFromSeed = Icarus.getKey
         . flip Icarus.generateKeyFromHardwareLedger pass
 
+    toPub (AXPrv xprv) = return . AXPub . CC.toXPub $ xprv
+    toPub (AXPub _) = Left "Input is already a public key."
+
     derive
         :: CC.DerivationScheme
-        -> XPrv
+        -> XPrvOrXPub
         -> DerivationIndex
-        -> Either String XPrv
-    derive scheme1Or2 k i =
-        return
+        -> Either String XPrvOrXPub
+    derive scheme1Or2 (AXPrv k) i =
+        return . AXPrv
             $ CC.deriveXPrv
                 scheme1Or2
                 pass
                 k
                 (unDerivationIndex i)
+    derive scheme1Or2 (AXPub k) i =
+        maybe (Left err) (return . AXPub)
+            $ CC.deriveXPub
+                scheme1Or2
+                k
+                (unDerivationIndex i)
+      where
+        err = mconcat
+            [ T.unpack (toText i)
+            , " is a hardened index. Public key derivation is only possible for"
+            , " soft indices. \nIf the index is correct, please use the"
```

Comment:
Small consistency tweak. This message uses both "indices" and "index" interchangeably, perhaps it's good to stick to one consistent wording otherwise people may start to wonder what's the difference between an indice and an index.

Reply:
I believe both `indices` and `indexes` are valid plural forms of `index`

- https://www.nasdaq.com/articles/indexes-or-indices-whats-the-deal-2016-05-12
- https://developer.apple.com/documentation/swift/collection/1641719-indices

Actually we are already using both even without this PR (though indices much less):

```
git grep "indices"
.travis.yml:  - rm -rf $HOME/.stack/indices/Hackage/00-index.tar $HOME/.stack/indices/Hackage/00-index.tar.gz
lib/cli/src/Cardano/CLI.hs:            , " soft indices. \nIf the index is correct, please use the"
lib/cli/test/unit/Cardano/CLISpec.hs:                    \derivation is only possible for soft indices. \nIf the\
lib/cli/test/unit/Cardano/CLISpec.hs:        it "derive . toPublic == toPublic . derive (for soft indices)"
lib/cli/test/unit/Cardano/CLISpec.hs:-- | For soft indices, public key derivation should be "equivalent" to private
lib/core/src/Cardano/Wallet/DB/Sqlite.hs:    -- Note: there are sorted indices on these columns.
lib/core/src/Cardano/Wallet/DB/Sqlite/TH.hs:-- Mapping of pool addresses to indices, and the slot
lib/core/src/Cardano/Wallet/Primitive/AddressDerivation/Byron.hs:    -- ^ The address derivation indices for the level of this key.
lib/core/src/Cardano/Wallet/Primitive/AddressDerivation/Byron.hs:-- | The hierarchical derivation indices for a given level/depth.
lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Random.hs:    -- that addresses will be _discovered_ from any and all account indices,

git grep "indexes"
lib/byron/test/data/cardano-node/genesis.yaml:  # Address indexes below are #501, #502, ..., #510
lib/byron/test/integration/Test/Integration/Byron/Scenario/CLI/Addresses.hs:        describe "CLI_ADDRESS_CREATE_07 - False indexes" $ do
lib/byron/test/unit/Cardano/Wallet/Byron/TransactionSpec.hs:    <*> scale (`mod` 3) arbitrary -- No need for a high indexes
lib/cli/src/Cardano/CLI.hs:                , "Please use \"H\" to denote hardened indexes. "
lib/cli/test/unit/Cardano/CLISpec.hs:                     \Please use \"H\" to denote hardened indexes. Did you \
lib/core-integration/src/Test/Integration/Scenario/API/Byron/Wallets.hs:    it "BYRON_RESTORE_08 - Icarus wallet with high indexes" $ \ctx -> do
lib/core-integration/src/Test/Integration/Scenario/API/Byron/Wallets.hs:        -- Special Icarus mnemonic where address indexes are all after the index
lib/core/src/Cardano/Byron/Codec/Cbor.hs:                , "expected two indexes but got: "
lib/core/src/Cardano/Wallet/DB/Sqlite.hs:                -- wallet backend was generating addresses indexes across the
lib/core/src/Cardano/Wallet/DB/Sqlite.hs:                -- We later changed that to instead use "hardened indexes". Yet,
lib/core/src/Cardano/Wallet/DB/Sqlite/TH.hs:-- Sequential address discovery scheme -- pending change indexes
lib/core/src/Cardano/Wallet/Primitive/AddressDerivation.hs:-- | Type of derivation that should be used with the given indexes.
lib/core/src/Cardano/Wallet/Primitive/AddressDerivation/Shelley.hs:-- indexes and soft (non-hardened) indexes for various level in a non-ambiguous
lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Sequential.hs:-- in order to discover addresses with indexes separated by possible huge gaps.
lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Sequential.hs:-- expected to be ordered in sequence (first indexes, first in the list).
lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Sequential.hs:-- | An ordered set of pending indexes. This keep track of indexes used
lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Sequential.hs:-- | An empty pending set of change indexes.
lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Sequential.hs:-- | Update the set of pending indexes by discarding every indexes _below_ the
lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Sequential.hs:-- to know from where we can generate new indexes.
lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Sequential.hs:-- | Construct a 'PendingIxs' from a list, ensuring that it is a set of indexes
lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Sequential.hs:-- | Get the next change index; If every available indexes have already been
lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Sequential.hs:-- taken, we'll rotate the pending set and re-use already provided indexes.
lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Sequential.hs:        <> indentF 4 ("Change indexes: " <> indentF 4 chgsF)
lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Sequential.hs:    -- | We pick indexes in sequence from the first known available index (i.e.
lib/core/src/Cardano/Wallet/Primitive/Types.hs:        -- their indexes. It matters also for serialization.
lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs:        -- NOTE: No need for a crazy high indexes
lib/core/test/unit/Cardano/Wallet/DB/Arbitrary.hs:        <*> scale (`mod` 3) arbitrary -- No need for a high indexes
lib/core/test/unit/Cardano/Wallet/DB/Arbitrary.hs:-- indexes. For example:
lib/core/test/unit/Cardano/Wallet/DB/Arbitrary.hs:-- - put checkpoint with pending indexes
lib/core/test/unit/Cardano/Wallet/DB/Arbitrary.hs:-- - put checkpoint with empty pending indexes
lib/core/test/unit/Cardano/Wallet/DB/Arbitrary.hs:-- currently, it'll also contains the pending indexes from the previously
lib/core/test/unit/Cardano/Wallet/Primitive/AddressDerivation/ByronSpec.hs:        it "Key derivation from seed works for various indexes" $
lib/core/test/unit/Cardano/Wallet/Primitive/AddressDerivation/ByronSpec.hs:        it "Key derivation from master key works for various indexes" $
lib/core/test/unit/Cardano/Wallet/Primitive/AddressDerivation/IcarusSpec.hs:        it "deriveAccountPrivateKey works for various indexes" $
lib/core/test/unit/Cardano/Wallet/Primitive/AddressDerivation/ShelleySpec.hs:        it "deriveAccountPrivateKey works for various indexes" $
lib/core/test/unit/Cardano/Wallet/Primitive/AddressDiscovery/RandomSpec.hs:    it "findUnusedPath: indexes are always in the 'hardened' realm" $
lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelectionSpec.hs:        <*> scale (`mod` 3) arbitrary -- No need for a high indexes
lib/core/test/unit/Cardano/Wallet/Primitive/FeeSpec.hs:        <*> scale (`mod` 3) arbitrary -- No need for a high indexes
lib/core/test/unit/Cardano/Wallet/Primitive/TypesSpec.hs:        <*> scale (`mod` 3) arbitrary -- No need for a crazy high indexes
lib/jormungandr/test/data/jormungandr/genesis.yaml:  # Address indexes below are #501, #502, ..., #510
lib/jormungandr/test/integration/Main.hs:-- outputs and the values of the input indexes.
```

ouroboros-network has more occurrences of indices than indexes though

😅 🔫